toolchain/parser/parser_handle_array_expression.cpp

```diff
@@ -0,0 +1,51 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lexer/token_kind.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+#include "toolchain/parser/parse_node_kind.h"
+#include "toolchain/parser/parser_context.h"
+#include "toolchain/parser/parser_state.h"
+
+namespace Carbon {
+
+auto ParserHandleArrayExpression(ParserContext& context) -> void {
+  auto state = context.PopState();
+  state.state = ParserState::ArrayExpressionSemi;
+  context.AddNode(ParseNodeKind::ArrayExpressionStart,
+                  context.ConsumeChecked(TokenKind::OpenSquareBracket),
+                  state.subtree_start, state.has_error);
+  context.PushState(state);
+  context.PushState(ParserState::Expression);
+}
+
+auto ParserHandleArrayExpressionSemi(ParserContext& context) -> void {
+  auto state = context.PopState();
+  auto semi = context.ConsumeIf(TokenKind::Semi);
+  if (!semi) {
+    context.AddNode(ParseNodeKind::ArrayExpressionSemi, *context.position(),
+                    state.subtree_start, true);
+    CARBON_DIAGNOSTIC(ExpectedArraySemi, Error,
+                      "Invalid array declaration. Expected Semi.");
```

Comment:
```suggestion
                      "Expected `;` in array type.");
```
Carbon developers seeing this error won't necessarily know what we mean by "Semi"; using `;` in diagnostics will be clearer. (Also, we don't know that this is an array declaration; it could be an array type appearing in some other context.)